### PR TITLE
Component extension support

### PR DESCRIPTION
**Pull request by roleyfoley**
## Description
Initial implementation of https://github.com/hamlet-io/engine/issues/1443 which aligns fragments ( now extensions ) with our dynamic macro invocation process.

instead of relying on a switch statement and a context variable assigned within a setup routine the context is now handled as part of the extension processing. This still maintains a globally assigned  `_context` variable but it has been moved into the core engine configuration where it is contents is managed through the extension. This maintains backwards compatibility with our existing fragment processing. 

Extensions also add some extra features over the existing fragments 
- Isolated macros reduce the processing time and only invokes the extension required for a component 
- Definition and documentation 
    - Extensions must be added to the configuration before they can be used, this includes defining a collection of properties that are used to validated the fragments
        - Aliases: Support using the same extension with multiple fragment names 
        - Component Type Support: Each extension can specify component types that it supports, if the component that is invoking the extension doesn't match the processing fails ( a wild card can be provided ) 
       - Descriptions: allow you to say what the extension does and to reflect this in documentation
- Inclusion as part of modules and plugins: extensions are loaded as part of our plugin process allowing them to be included in plugins which are loaded as required and don't require a CMDB 

## Motivation and Context
Covered in https://github.com/hamlet-io/engine/issues/1443 but essentially the main motivation for this is to support adding extensions in plugins 

## How Has This Been Tested?
Tested locally 

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves the structure or operation of the implementation)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Followup Actions
- [x] None

## Checklist:
- [ ] My change requires a change to the [documentation](https://github.com/hamlet-io/docs).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [X] None of the above.

### CODE REVIEW

File: engine/extension.ftl [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+            ]]
+
+        [#local extensionMacro = getFirstDefinedDirective(extensionMacroOptions)]
+        [#if extensionMacro?has_content ]
+
+            [#-- Validate the extenion usage --]
+            [#if ! extensionDetails?has_content ]
+                [@fatal
+                    message="Could not find extension details"
+                    detail="An extension function was found by could not find its configuration details"
```
Comment (ml019) [inline engine/extension.ftl:190]: typo

File: engine/extension.ftl [hunk 2]
```diff
@@ -0,0 +1,251 @@
+[#ftl]
+
+[#-- Component Extensions --]
+[#-- Allows for the modification of a components state and its setup routine by users --]
+[#-- Generally used for advanced configuration scenarios including programitc settings and adding specialised resources --]
```
Comment (ml019) [inline engine/extension.ftl:5]: typo

File: engine/extension.ftl [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+[#-- Extensions are defined as part plugin but can also be added to users CMDBs as an embeded extension --]
```
Comment (ml019) [inline engine/extension.ftl:6]: typo

**Review by ml019 — CHANGES_REQUESTED** [review]
Very elegant. Ability to apply multiple extensions an a mix of plugin provided and local will be good.

File: engine/extension.ftl (hunk 1 above)
Comment (roleyfoley) [reply engine/extension.ftl]: Fixed

File: engine/extension.ftl (hunk 3 above)
Comment (roleyfoley) [reply engine/extension.ftl]: Fixed

**Comment by roleyfoley** [comment]
@ml019 Typos fixed up - you can also use the suggest changes option in the comment for stuff like that ( https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/commenting-on-a-pull-request ) 

In a PR comment box the small piece of paper with the + sign lets you suggest udpates which can be merged straight from the github console

File: engine/bootstrap.ftl [hunk 4]
```diff
@@ -34,6 +34,9 @@
 [#-- Module Loading --]
 [#include "module.ftl" ]
 
+[#-- Extension loading --]
+[#include "extension.ftl" ]
```
Comment (ml019) [inline engine/bootstrap.ftl:38]:
```suggestion
[#include fred.ftl]
```

**Comment by ml019** [comment]
@roleyfoley thanks for the tip. Will help in future.

**Pull request merged**